A Moral Manifest

hooksta4 opened the pull request:
A Moral Manifest (BEASTMAN HEADS)

<!-- place 'x' mark between square [] brackets to affirm: -->
**_I affirm:_**
- [x] that I agree to Project Topaz's [Limited Contributor License Agreement](http://project-topaz.com/blob/release/CONTRIBUTOR_AGREEMENT.md), as written on this date
- [x] that I've _tested my code_ since the last commit in the PR, and will test after any later commits

hooksta4 committed:
A Moral Manifest

A Moral Manifest (BEASTMAN HEADS)

Omnione commented on scripts/zones/Altar_Room/IDs.lua:
```diff
@@ -19,6 +19,17 @@ zones[tpz.zone.ALTAR_ROOM] =
+        GOO_PAKE_THE_BLOODHOUND = 17399817,
+
+
     },
```
Alignment is a bit off here.

Omnione commented on scripts/zones/Altar_Room/mobs/Yagudo_Avatar.lua:
```diff
@@ -0,0 +1,27 @@
+function onMobSpawn(mob)
+end
+
+function onMobDeath(mob, player, isKiller)
+    if  player:getQuestStatus(OTHER_AREAS_LOG,tpz.quest.id.otherAreas.A_MORAL_MANIFEST) == QUEST_ACCEPTED and player:getCharVar("moral") == 5 then
+        player:setCharVar("moral", 6)
+        player:delKeyItem(tpz.ki.VAULT_QUIPUS);
+    end
+    for i = ID.mob.YAGUDO_AVATAR + 1, ID.mob.YAGUDO_AVATAR + 8 do
+        DespawnMob(i)
+    end
+end
+
+function onMobDespawn(mob)
+    for i = ID.mob.YAGUDO_AVATAR + 1, ID.mob.YAGUDO_AVATAR + 8 do
+        DespawnMob(i)
+    end
+end
```
You might want a check to see if these mobs are alive before despawning them to avoid errors.
something like:
```
if not GetMobByID(i):isDead() then
    DespawnMob(i)
end
```

someone else might have a better method to add, this is just a quick example.

TeoTwawki replied:
This actually shouldn't be a problem, and if it is we should make the core not let that be a problem. If I tell an entity to despawn that is already in the process of doing so it should just do nothing (as soon as they die they code sends them to the despawn state).


p.s.   
there is(was?) a rare known related bug where if you despawn an entity while it is actively fighting and its melee round completed after the despawn command hits it gets stuck. I dunno if that was ever fixed, but thats also a thing to handle core side outside of this pr if it wasn't

Omnione commented on scripts/zones/Altar_Room/npcs/Stone_Lid.lua:
```diff
@@ -0,0 +1,57 @@
+
+function onEventFinish(player,csid,option)
+    if csid == 48 then
+        local mobs =
+        {
+            ID.mob.YAGUDO_AVATAR,
+            ID.mob.DUU_MASA_THE_ONECUT,
+            ID.mob.FEE_JUGE_THE_RAMFIST,
+            ID.mob.GOO_PAKE_THE_BLOODHOUND,
+            ID.mob.KEE_TAW_THE_NIGHTINGALE,
+            ID.mob.LAA_YAKU_THE_AUSTERE,
+			ID.mob.POO_YOZO_THE_BABBLER,
+        }
```
just a small alignment off on one line, nothing major.

Omnione commented on scripts/zones/Windurst_Woods/npcs/Ponono.lua:
```diff
@@ -11,34 +14,58 @@ require("scripts/globals/status")
         player:setSkillRank(tpz.skill.CLOTHCRAFT,newRank)
         player:startEvent(10012,0,0,0,0,newRank)
     end
+    if moralmanifest == QUEST_ACCEPTED and player:getCharVar("moral") == 2 then
+        if (trade:hasItemQty(828,1) and trade:hasItemQty (830, 1) and tradeGil == 10000) then -- Trade Velvet Cloth, Rainbow Cloth and 10k
+            player:tradeComplete()
```
I think we are trying to get everyone to start using the **npcUtil.tradeHas** method of checking items now if possible, to avoid issues,
A reveiwer can elaborate more on this if possible, but the function is found in the global npc_util.lua,
you should be able to search other scripts using this style of trading, again i may be incorrect, i am just pointing out potential issues for reviewers to look through, dont feel this is an issue unless others have asked to have this changed.

Omnione reviewed (commented):
Other than the alignment thing and potential trade issue, its looking good, just needs a more in depth review.

hooksta4 commented:
Thanks! Will look into that trade function and clean up a bit. Always miss a few

hooksta4 committed:
Fixes Round 1

Some npcUtil and format fixes.

Omnione reviewed (approved):
Nice one, looks good

Omnione commented:
Looks like travis failed, does not appear to be code related.

hooksta4 commented:
what could it have been?

Omnione commented:
It happens sometimes, im sure it will get looked at when someone is available to kick into gear again

ibm2431 commented:
> scripts/zones/Windurst_Woods/npcs/Ponono.lua:24:45: (E011) expected '}' near '{'

TeoTwawki commented on scripts/zones/Windurst_Woods/npcs/Ponono.lua:
```diff
@@ -11,34 +14,57 @@ require("scripts/globals/status")
         player:setSkillRank(tpz.skill.CLOTHCRAFT,newRank)
         player:startEvent(10012,0,0,0,0,newRank)
     end
+    if moralmanifest == QUEST_ACCEPTED and player:getCharVar("moral") == 2 then
```
both conditions can't be triggered in one trade, so use `elseif`

hooksta4 replied:
Not sure what you mean by this, maybe help me out a bit - this specific is under trigger vice trade. Through test I have found that this works based on if they are in the guild or not. Maybe someone else could help test and see if I am not seeing something correctly?

TeoTwawki replied:
```lua
if stuff then
    code
elseif morestuff then
    more code
end
```
vs
```lua
if stuff then
   code
end
if stuff then
   code
end
```
not all mistakes produce an error.

TeoTwawki replied:
this still needs fixed

TeoTwawki replied:
pardon, I commented at wrong place

TeoTwawki replied:
the way its structured now, your code here potentially tries to start an event after an event may have already started.

zach2good replied:
Agree with this ^

TeoTwawki commented on scripts/zones/Windurst_Woods/npcs/Ponono.lua:
```diff
@@ -11,34 +14,57 @@ require("scripts/globals/status")
+    elseif player:getCharVar("moral") == 3 and player:getLocalVar("moralZone") == 0 and player:getCharVar("moralWait") <= os.time() then
+        player:startEvent(705)
+    else player:startEvent(10011,testItem,getNewRank,30,guildMember,44,0,0,0)
+	end
```
indention fubar plus missing newline after the `else`

TeoTwawki replied:
still need that newline

```lua
    if moralmanifest == QUEST_ACCEPTED and player:getCharVar("moral") == 1 then
        player:startEvent(700)
    elseif moralmanifest == QUEST_COMPLETE or moralmanifest == QUEST_ACCEPTED and player:getCharVar("moral") >=4  then
        player:startEvent(704)
    elseif player:getCharVar("moral") == 3 and player:getLocalVar("moralZone") == 0 and player:getCharVar("moralWait") <= os.time() then
        player:startEvent(705)
    else
        player:startEvent(10011,testItem,getNewRank,30,guildMember,44,0,0,0)
	end

zach2good replied:
Agreed ^

TeoTwawki commented on scripts/zones/Windurst_Woods/npcs/Ponono.lua:
```diff
@@ -11,34 +14,57 @@ require("scripts/globals/status")
         player:startEvent(10012,0,0,0,0,newRank)
     end
+    if moralmanifest == QUEST_ACCEPTED and player:getCharVar("moral") == 2 then
+        if npcUtil.tradeHas(trade, {828, 830{"gil", 10000}}) then -- Trade Velvet Cloth, Rainbow Cloth and 10k
```
bracing error

TeoTwawki commented on scripts/zones/Altar_Room/IDs.lua:
```diff
@@ -19,6 +19,17 @@ zones[tpz.zone.ALTAR_ROOM] =
     },
     mob =
     {
+	    YAGUDO_AVATAR           = 17399809,
```
misaligned plus got some extra newlines below before the closing brace

TeoTwawki commented on scripts/zones/Windurst_Woods/npcs/Ponono.lua:
```diff
@@ -11,34 +14,57 @@ require("scripts/globals/status")
         player:startEvent(10012,0,0,0,0,newRank)
     end
+    if moralmanifest == QUEST_ACCEPTED and player:getCharVar("moral") == 2 then
+        if npcUtil.tradeHas(trade, {828, 830{"gil", 10000}}) then -- Trade Velvet Cloth, Rainbow Cloth and 10k
```
this line errors, misplaced/missing comma _I think_

hooksta4 replied:
Can you give an example of npctil with items and Gil? Tried to find but couldn't.

Omnione replied:
I think what you have is right, but as Teo suggested, you have a comma missing
```if npcUtil.tradeHas(trade, {828, 830,{"gil", 10000}}) then```
it should be like that i belive.

Omnione replied:
let us know if that worked

TeoTwawki replied:
This is why testing again after you change something is important. If you have difficulty getting it to work when testing, hit folks up for help - discord is usually pretty active with ppl willing to assist

>     * [x]  that I've _tested my code_ since the last commit in the PR, and will test after any later commits

hooksta4 replied:
@TeoTwawki  I usually do - Not sure why I didn't catch it... might have not saved it. I will ensure, moving forward, I double check.

hooksta4 replied:
So I just ran through it entirely again and didn't have any hang ups. ( I did do give myself item 15202 and set moralWait to 0)

hooksta4 committed:
More fixes

Used @TeoTwawki and @Omnione  recommendations and patched what I could. Please review.

59blargedy commented:
@project-topaz/developer  @project-topaz/reviewer  would love an additional approval on here so this can be merged and tested. thanks for the help

hooksta4 commented:
Once this one gets tested and merged I have the other 3 in queue. In principle, same coding. Thanks for everyone's hard work! @59blargedy @TeoTwawki @ibm2431

zach2good commented on scripts/zones/Windurst_Woods/npcs/Ponono.lua:
```diff
@@ -3,6 +3,9 @@
 --  NPC: Ponono
 -- Type: Clothcraft Guild Master
 -- !pos -38.243 -2.25 -120.954 241
+-- TODO Allow players to purchase additional Cuttings
+-- TODO Allow players to cancel quest 
+-- TODO Requre check for other 3 quests An Understanding General, A Generous General, 
```
Once this gets merged, please make issues for these TODOs 👍

zach2good commented on scripts/zones/Windurst_Woods/npcs/Ponono.lua:
```diff
@@ -11,34 +14,57 @@ require("scripts/globals/status")
     local craftSkill = player:getSkillLevel(tpz.skill.CLOTHCRAFT)
     local testItem = getTestItem(player,npc,tpz.skill.CLOTHCRAFT)
     local guildMember = isGuildMember(player,3)
+    local moralmanifest = player:getQuestStatus(OTHER_AREAS_LOG,tpz.quest.id.otherAreas.A_MORAL_MANIFEST)
```
Very minor styling in line with the other local vars:
`moralmanifest` -> `moralManifest`

zach2good commented on scripts/zones/Altar_Room/mobs/Duu_Masa_The_Onecut.lua:
```diff
@@ -0,0 +1,9 @@
+-----------------------------------
+
+function onMobDeath(mob, player, isKiller)
+end
```
Newline at the ends of new files please :)

zach2good reviewed (changes requested):
Pretty much just small styling issues, and that one `if + if` vs `if + if else` thing Teo mentioned. Looks really good otherwise, nice work 👍

zach2good commented:
I'm gonna merge this to a WIP branch and handle the changes since they're small :)

hooksta4 commented:
Thank you! I've been swamped at work and haven't been able to get on much.

zach2good commented:
No problem, feel free to pick up where you left off whenever you have time and feel ready :)
